src/Storages/HDFS/StorageHDFS.cpp

```diff
@@ -218,26 +210,36 @@ ColumnsDescription StorageHDFS::getTableStructureFromData(
 
     ReadBufferIterator read_buffer_iterator
         = [&, my_uri_without_path = uri_without_path, it = paths_with_info.begin(), first = true](
-              ColumnsDescription & columns) mutable -> std::unique_ptr<ReadBuffer>
+              ColumnsDescription &) mutable -> std::unique_ptr<ReadBuffer>
     {
-        if (it == paths_with_info.end())
+        PathWithInfo path_with_info;
+        std::unique_ptr<ReadBuffer> buf;
+        while (true)
         {
-            if (first)
-                throw Exception(ErrorCodes::CANNOT_EXTRACT_TABLE_STRUCTURE,
-                                "Cannot extract table structure from {} format file, because all files are empty. "
-                                "You must specify table structure manually", format);
-            return nullptr;
-        }
+            if (it == paths_with_info.end())
+            {
+                if (first)
+                    throw Exception(ErrorCodes::CANNOT_EXTRACT_TABLE_STRUCTURE,
+                                    "Cannot extract table structure from {} format file, because all files are empty. "
+                                    "You must specify table structure manually", format);
+                return nullptr;
+            }
+
+            path_with_info = *it++;
+            if (ctx->getSettingsRef().hdfs_skip_empty_files && path_with_info.info && path_with_info.info->size == 0)
+                continue;
 
-        auto path_with_info = *it++;
-        if (ctx->getSettingsRef().hdfs_skip_empty_files && path_with_info.info && path_with_info.info->size == 0)
-            return read_buffer_iterator(columns);
+            auto compression = chooseCompressionMethod(path_with_info.path, compression_method);
+            auto impl = std::make_unique<ReadBufferFromHDFS>(my_uri_without_path, path_with_info.path, ctx->getGlobalContext()->getConfigRef(), ctx->getReadSettings());
+            const Int64 zstd_window_log_max = ctx->getSettingsRef().zstd_window_log_max;
+            buf = wrapReadBufferWithCompressionMethod(std::move(impl), compression, static_cast<int>(zstd_window_log_max));
 
-        first = false;
-        auto compression = chooseCompressionMethod(path_with_info.path, compression_method);
-        auto impl = std::make_unique<ReadBufferFromHDFS>(my_uri_without_path, path_with_info.path, ctx->getGlobalContext()->getConfigRef(), ctx->getReadSettings());
-        const Int64 zstd_window_log_max = ctx->getSettingsRef().zstd_window_log_max;
-        return wrapReadBufferWithCompressionMethod(std::move(impl), compression, static_cast<int>(zstd_window_log_max));
+            if (!ctx->getSettingsRef().hdfs_skip_empty_files || !buf->eof())
```

Comment:
Need to check `eof()` *before* wrapping. Otherwise the decompressor will error out on empty input.